internal/controller/pkg/revision/hook.go

```diff
@@ -215,6 +228,128 @@ func (h *ConfigurationHooks) Post(context.Context, runtime.Object, v1.PackageRev
 	return nil
 }
 
+// FunctionHooks performs operations for a function package that requires a
+// controller before and after the revision establishes objects.
+type FunctionHooks struct {
+	client         resource.ClientApplicator
+	namespace      string
+	serviceAccount string
+}
+
+// NewFunctionHooks creates a new FunctionHooks.
+func NewFunctionHooks(client resource.ClientApplicator, namespace, serviceAccount string) *FunctionHooks {
+	return &FunctionHooks{
+		client:         client,
+		namespace:      namespace,
+		serviceAccount: serviceAccount,
+	}
+}
+
+// Pre cleans up a packaged controller and service account if the revision is
+// inactive.
+func (h *FunctionHooks) Pre(ctx context.Context, pkg runtime.Object, pr v1.PackageRevision) error {
+	fo, _ := xpkg.TryConvert(pkg, &pkgmetav1alpha1.Function{})
+	pkgFunction, ok := fo.(*pkgmetav1alpha1.Function)
+	if !ok {
+		return errors.New(errNotFunction)
+	}
+
+	// TODO(ezgidemirel): update any status fields relevant to package revisions.
+
+	// Do not clean up SA and controller if revision is not inactive.
+	if pr.GetDesiredState() != v1.PackageRevisionInactive {
+		return nil
+	}
+
+	// NOTE(hasheddan): we avoid fetching pull secrets and controller config as
+	// they aren't needed to delete Deployment, ServiceAccount, and Service.
+	s, d, _, _ := buildFunctionDeployment(pkgFunction, pr, nil, h.namespace, []corev1.LocalObjectReference{})
```

Comment:
nit: maybe update the comment above and explain why service and server secret are not deleted here (which one could expect)

Reply:
I can do it in the following PR where I'll start creating resources per packages, not revisions.

Reply:
why not in this one? it's just a comment 😄